src/Sylius/Component/Core/Statistics/Provider/SalesStatisticsProvider.php

```diff
@@ -63,4 +111,52 @@ private function getPeriodFormat(string $intervalType): string
 
         return $this->formatsMap[$intervalType];
     }
+
+    /**
+     * @param array<array{period: \DateTimeInterface, ...}> $firstArray
+     * @param array<array{period: \DateTimeInterface, ...}> $secondArray
+     *
+     * @return array<array{period: \DateTimeInterface, ...}>
+     */
+    private function mergeArraysByPeriod(array $firstArray, array $secondArray): array
+    {
+        $indexByPeriod = function(array $items) {
+            $result = [];
+            foreach ($items as $item) {
+                if (!isset($item['period']) || !$item['period'] instanceof \DateTimeInterface) {
+                    continue;
+                }
+                $key = $item['period']->format('c');
+                $result[$key] = $item;
+            }
+
+            return $result;
+        };
+
+        $indexedFirstArray = $indexByPeriod($firstArray);
+        $indexedSecondArray = $indexByPeriod($secondArray);
+
+        $allKeys = array_unique(array_merge(array_keys($indexedFirstArray), array_keys($indexedSecondArray)));
+
+        $result = [];
+        foreach ($allKeys as $key) {
+            $itemFromFirstArray = $indexedFirstArray[$key] ?? [];
+            $itemFromSecondArray = $indexedSecondArray[$key] ?? [];
+
+            $period = $itemFromFirstArray['period'] ?? $itemFromSecondArray['period'] ?? null;
+            if (!$period instanceof \DateTimeInterface) {
+                continue;
+            }
+
+            unset($itemFromFirstArray['period'], $itemFromSecondArray['period']);
+
+            $result[] = array_merge(
+                ['period' => $period],
+                $itemFromFirstArray,
+                $itemFromSecondArray,
+            );
+        }
```

Comment:
I have a feeling this would work great as a generator, and again, potentially removing the merge